src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -411,17 +411,19 @@ private ValueTask SendFrameLockAcquiredNonCancelableAsync(MessageOpcode opcode,
             // If we get here, the cancellation token is not cancelable so we don't have to worry about it,
             // and we own the semaphore, so we don't need to asynchronously wait for it.
             ValueTask writeTask = default;
+            Task flushTask;
             bool releaseSendBufferAndSemaphore = true;
             try
             {
                 // Write the payload synchronously to the buffer, then write that buffer out to the network.
                 int sendBytes = WriteFrameToSendBuffer(opcode, endOfMessage, disableCompression, payloadBuffer.Span);
                 writeTask = _stream.WriteAsync(new ReadOnlyMemory<byte>(_sendBuffer, 0, sendBytes));
+                flushTask = _stream.FlushAsync();
```

Comment:
Flush should always be triggered *after* the write, to be sure that we will be flushing whatever we just wrote. Now you have two parallel tasks, so Flush could execute earlier and data from your write would not be sent.

Reply:
Moved after write completion